### PR TITLE
refactor(engine): Wire-up basic Python protocol run logic

### LINKED ISSUE #7844: Protocol Engine milestone: Upload and run basic Python protocols
# User story

As a 3rd-party developer, I want to programmatically upload and run a basic Python APIv3 protocol on the OT-2.

# Acceptance criteria

* [ ] Using an HTTP API, I can upload a "basic" Python APIv3 protocol to the OT-2.
  * "Basic" means single-file, excluding modules, 8-Channel pipettes, and so on.
* [ ] Using an HTTP API, I can execute that protocol and make the robot move.
  * Execution must happen through Protocol Engine.
  * For the purposes of this story, we're going straight to execution. We don't care about calibration, metadata inspection, requirements inspection, or run log generation.
* [ ] If a protocol is already running, I'm disallowed from starting a new one.
	 * Bonus points: also don't let me start calibration or other stuff that would interfere.

We've punted running APIv2 protocols out of this sprint. For now, we can assume that all protocols coming through the new HTTP API are APIv3.

# Implementation suggestions

It could work like this (courtesy @sanni-t):

* An endpoint like `POST /protocols` takes the protocol file as form data.
* The robot-server, upon receiving the request with the file, sends the file to Python Protocol Runner, which may simulate and extract metadata (or these may be stubbed).
* Appropriate response to robot-server should include:
  * Metadata w/ Labware, modules, pipettes (empty fields)
  * Simulation result & warnings? (empty fields)
  * Handles to get:
    * Labware definitions
    * Protocol steps (Run log)
    * Create session
* A `create_session` endpoint uses the file runner to build (get?) protocol context and fetches the PE client to execute commands using PE. (File runner details in separate PR) 

See also #7647.

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR implements a very basic `PythonFileRunner` so the the ProtocolEngine-based `/sessions` API can run Python protocols!

Closes #7844 

## Changelog

- Implement `PythonFileRunner` that runs Python protocols

## Review requests

As usual, make sure your robot / dev server has the `enableProtocolEngine` feature-flag enabled. Use the Opentrons App to set this on your robot, or launch your dev server with `make -C robot-server dev OT_API_FF_enableProtocolEngine=1`

This PR adds a test Python protocol: [`api/tests/opentrons/data/testosaur_v3.py`](https://github.com/Opentrons/opentrons/pull/8032/files#diff-f103c459caec0844c858b26c9d905a64720302ea6f9a4c543e48eb03796dd19b). You can run this Python protocol using the [Postman testing collection](https://gist.github.com/mcous/7dde93b6ea83753b86b0d84b40ba07f6).

**Note**: due to required changes in the engine-based Python Protocol API as well as issue #8033, this protocol will fail at the first attempt to do any sort of well interaction. This is ok for this PR! Make sure that the error is logged in the server, though.

## Risk assessment

Risk here is due to this being foundational work for future production logic. For now, this is feature flagged

### CODE REVIEW

File: api/tests/opentrons/file_runner/smoke_tests/test_python.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+@pytest.fixture
+def python_protocol_file(tmp_path: Path) -> ProtocolFile:
+    """Get an on-disk, minimal Python protocol fixture."""
+    file_path = tmp_path / "protocol-name.py"
+    file_path.write_text(
+        textwrap.dedent(
+            """
+            # my protocol
+            metadata = {
+                "apiVersion": "3.0",
```
Comment (sanni-t) [inline api/tests/opentrons/file_runner/smoke_tests/test_python.py]:
```suggestion
                "apiLevel": "3.0",
```

File: robot-server/robot_server/service/task_runner.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+"""Background task management.
+
+This mosule is mostly a thin wrapper around fastapi.BackgroundTasks
```
Comment (sanni-t) [inline robot-server/robot_server/service/task_runner.py]:
```suggestion
This module is mostly a thin wrapper around fastapi.BackgroundTasks
```

File: robot-server/robot_server/service/task_runner.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        self._background_tasks = background_tasks
+
+    def run(self, func: TaskFunc) -> None:
+        """Run a function in the background.
+
+        Will log when the function completes, including any error
+        that may occur.
+
+        Arguments:
+            func: A argumentless, None-returing function to run the in
```
Comment (sanni-t) [inline robot-server/robot_server/service/task_runner.py]:
```suggestion
            func: An argumentless, None-returing function to run in
```

File: robot-server/tests/sessions/conftest.py [hunk 4]
```diff
@@ -124,7 +131,7 @@ def python_protocol_file(tmp_path: Path) -> Path:
             """
             # my protocol
             metadata = {
-                "apiVersion": "3.0"
+                "apiVersion": "3.0",
```
Comment (sanni-t) [inline robot-server/tests/sessions/conftest.py]:
```suggestion
                "apiLevel": "3.0",
```

**Review by sanni-t — APPROVED** [review]
🏃🙌🎉

File: api/src/opentrons/file_runner/json_file_runner.py [hunk 5]
```diff
@@ -41,8 +41,15 @@ def load(self) -> None:
         for cmd in translated_items:
             self._protocol_engine.add_command(cmd)
 
+    async def run(self) -> None:
+        """Run the protocol to completion."""
+        # TODO(mc, 2021-06-30): this will not work with pause/resume. Rework
+        # so that queue worker has `wait_for_done` method, instead.
```
Comment (SyntaxColoring) [inline api/src/opentrons/file_runner/json_file_runner.py:47]: Yeah.

Or even an old-school condition variable. Like:

1. wait for any "state change"
2. then, move on if the current state is `completed` or `fatally_errored`. Else, wait for the next state change and check again.

**Pull request merged**